crates/executor/src/estimate.rs

```diff
@@ -13,54 +19,391 @@ pub fn estimate(
 
     let (mut state, block_context) = execution_state.starknet_state()?;
 
-    let mut fees = Vec::with_capacity(transactions.len());
-    for (transaction_idx, transaction) in transactions.into_iter().enumerate() {
-        let _span = tracing::debug_span!("estimate", transaction_hash=%super::transaction::transaction_hash(&transaction), %block_number, %transaction_idx).entered();
+    transactions
+        .into_iter()
+        .enumerate()
+        .map(|(tx_index, mut tx)| {
+            let _span = tracing::debug_span!(
+                "estimate",
+                block_number = %block_number,
+                transaction_hash = %transaction_hash(&tx),
+                transaction_index = %tx_index
+            )
+            .entered();
 
-        let fee_type = super::transaction::fee_type(&transaction);
-        let gas_vector_computation_mode =
-            super::transaction::gas_vector_computation_mode(&transaction);
-        let minimal_l1_gas_amount_vector = match &transaction {
+            let gas_vector_computation_mode = super::transaction::gas_vector_computation_mode(&tx);
+            let tx_info = match gas_vector_computation_mode {
+                GasVectorComputationMode::NoL2Gas => {
+                    execute_transaction(&tx, tx_index, &mut state, &block_context)?
+                }
+                GasVectorComputationMode::All => find_l2_gas_limit_and_execute_transaction(
+                    &mut tx,
+                    tx_index,
+                    &mut state,
+                    &block_context,
+                )?,
```

Comment:
This is slightly unrelated, but I think `gas_vector_computation_mode` is maybe not the right way to decide if we need to do a binary search?

Like: if you're sending a tx with all three resource bounds specified then L2 gas will be used. Still, if classes involved have not been compiled as Sierra 1.7 then L2 gas costs will still be based on VM resources and not on L2 gas accounting -- thus no redeposits are done.

I wonder what's the correct way to find out if this is indeed the case...